File: SceneViewSwift/Sources/SceneViewSwift/Audio/AudioFalloff.swift
#if os(iOS) || os(macOS) || os(visionOS)
import Foundation

/// Distance-attenuation model applied to a ``SpatialAudioNode``.
///
/// Mirrors Android's `io.github.sceneview.audio.AudioFalloff` sealed hierarchy verbatim so
/// the same `(refDistance, maxDistance)` inputs produce identical gain curves on every
/// platform. See the Android tests
/// (`sceneview/src/test/java/io/github/sceneview/audio/AudioFalloffTest.kt`) for the
/// golden table that every implementation must match — Swift `Self.gain(for:distance:)`
/// is the verbatim port and the Swift unit tests pin the same numbers.
public enum AudioFalloff: Sendable {

    /// Inverse distance attenuation (WebAudio `"inverse"` model). Physically realistic;
    /// the default for ``SpatialAudioNode``.
    case inverse(refDistance: Float = 1, maxDistance: Float = 100, rolloffFactor: Float = 1)

    /// Linear attenuation — gain reaches `0` exactly at `maxDistance`.
    case linear(refDistance: Float = 1, maxDistance: Float = 100)

    /// No distance attenuation — gain is always `1.0`.
    case none

    /// Computes the linear gain (`0...1`) for `falloff` at the given source-listener
    /// `distance` (meters). Internal helper used by the iOS RealityKit playback path so
    /// the curves match Android `AudioFalloff.gainFor`.
    public static func gain(for falloff: AudioFalloff, distance: Float) -> Float {
        // NaN / negative distances become 0 — keep the gain in `[0, 1]` even if a
        // malformed frame produces garbage coordinates.
        let d = (distance.isFinite && distance >= 0) ? distance : 0
        switch falloff {
        case .none:
            return 1
        case .inverse(let ref, let max, let rolloff):
            let clamped = Swift.min(d, max)
            if clamped <= ref { return 1 }
            let safeRef = Swift.max(ref, 1e-4)
            let gain = safeRef / (safeRef + rolloff * (clamped - safeRef))
            return Swift.min(Swift.max(gain, 0), 1)
        case .linear(let ref, let max):
            if max <= ref {
                return d < ref ? 1 : 0
            }
            let clamped = Swift.min(Swift.max(d, ref), max)
            let gain = 1 - (clamped - ref) / (max - ref)
            return Swift.min(Swift.max(gain, 0), 1)
        }
    }
}
#endif // os(iOS) || os(macOS) || os(visionOS)

File: sceneview/src/main/java/io/github/sceneview/audio/AudioFalloff.kt
package io.github.sceneview.audio

/**
 * Distance-attenuation model applied to a [SpatialAudioNode].
 *
 * The library evaluates the model on every frame against the current camera-to-source distance
 * and feeds the resulting linear gain into the audio backend (Android `Spatializer` /
 * `AudioTrack`, or the `MediaPlayer` fallback's `setVolume(left, right)`).
 *
 * The three variants mirror the standard Web Audio `PannerNode.distanceModel` settings — see the
 * Web Audio spec (https://webaudio.github.io/web-audio-api/#distance-effects) — so behaviour is
 * consistent across Android / iOS / Web.
 */
sealed interface AudioFalloff {

    /**
     * Inverse distance attenuation:
     *
     * `gain = refDistance / (refDistance + rolloffFactor * (clampedDistance - refDistance))`
     *
     * where `clampedDistance = min(distance, maxDistance)`. Below [refDistance] the gain is
     * `1.0`; above [maxDistance] the gain is clamped at the value computed at [maxDistance].
     * This is the WebAudio `"inverse"` model — physically the most realistic of the three.
     *
     * @param refDistance   Distance (m) at which gain is `1.0`. Default `1f`.
     * @param maxDistance   Distance (m) beyond which the gain stops decreasing further.
     * @param rolloffFactor How quickly gain drops once past [refDistance]. `1f` matches the
     *                      physical inverse-square style curve; higher values fall off faster.
     */
    data class Inverse(
        val refDistance: Float = 1f,
        val maxDistance: Float = 100f,
        val rolloffFactor: Float = 1f
    ) : AudioFalloff

    /**
     * Linear distance attenuation:
     *
     * `gain = 1 - (clamp(distance, refDistance, maxDistance) - refDistance) / (maxDistance - refDistance)`
     *
     * Reaches `0.0` exactly at [maxDistance]. Equivalent to the WebAudio `"linear"` model.
     *
     * @param refDistance Distance (m) at which gain is `1.0`. Default `1f`.
     * @param maxDistance Distance (m) at which gain hits `0.0`. Default `100f`.
     */
    data class Linear(
        val refDistance: Float = 1f,
        val maxDistance: Float = 100f
    ) : AudioFalloff

    /**
     * Disable distance attenuation. Always returns `1.0` regardless of source-listener distance.
     * Useful for ambient or non-positional cues.
     */
    object None : AudioFalloff

    companion object {
        /**
         * Computes the linear gain (`0f..1f`) for [falloff] at the given source-listener
         * [distance] (meters). Internal helper used by every backend so the curves match
         * across Android Spatializer, MediaPlayer L/R pan, iOS PHASE, and Web PannerNode.
         */
        fun gainFor(falloff: AudioFalloff, distance: Float): Float {
            // Safety: NaN / negative distances become 0 — keeps the gain in `[0, 1]`
            // even if a malformed frame produces garbage coordinates (Float3 NaN
            // cascades through every quaternion / matrix op upstream and would
            // otherwise nuke the audio bus volume too).
            val d = if (distance.isFinite() && distance >= 0f) distance else 0f
            return when (falloff) {
                is None -> 1f
                is Inverse -> {
                    val clamped = minOf(d, falloff.maxDistance)
                    // Below refDistance everything is at unity — same convention as
                    // OpenAL / WebAudio inverse model.
                    if (clamped <= falloff.refDistance) 1f
                    else {
                        val ref = falloff.refDistance.coerceAtLeast(1e-4f)
                        val gain = ref / (ref + falloff.rolloffFactor *
                                (clamped - ref))
                        gain.coerceIn(0f, 1f)
                    }
                }
                is Linear -> {
                    val ref = falloff.refDistance
                    val max = falloff.maxDistance
                    if (max <= ref) {
                        // Degenerate range — fall back to step at refDistance so the
                        // demo doesn't divide by zero.
                        if (d < ref) 1f else 0f
                    } else {
                        val clamped = d.coerceIn(ref, max)
                        (1f - (clamped - ref) / (max - ref)).coerceIn(0f, 1f)
                    }
                }
            }
        }
    }
}
